### PR TITLE
8283541: Add Statical counters and some comments in PhaseStringOpts

### PR DESCRIPTION
**Pull request by navyxliu**
Add 3 counters for `OptimizeStringConcat`.  Total is the total number of StringBuilder/Buffer.toString() encountered.  if it matches, it is either merged or replaced. 

1. For each StringConcat, increase `total` counter.
2. merged: this phase realizes that it can coalesce 2 StringConcats, or 
3. replaced: this phase replace a StringConcat with a new String. 

In the following example, javac encounters 79 StringConcats, 4 of them are merged with their successors.  41 have been replaced.  The remaining 34 are mismatched in `build_candidate`. 
```
$./build/linux-x86_64-server-fastdebug/images/jdk/bin/javac -J-Xcomp -J-XX:+PrintOptoStatistics

--- Compiler Statistics ---
Methods seen: 13873  Methods parsed: 13873  Nodes created: 3597636
Blocks parsed: 42441  Blocks seen: 46403
50086 original NULL checks - 41382 elided (82%); optimizer leaves 13545,
3671 made implicit (27%)
36 implicit null exceptions at runtime
StringConcat:   41/   4/  79(replaced/merged/total)
...
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8283541](https://bugs.openjdk.java.net/browse/JDK-8283541): Add Statical counters and some comments in PhaseStringOpts

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to 62a5f9ff429bc6aa9a0a50f1155f283ec803d794
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to 62a5f9ff429bc6aa9a0a50f1155f283ec803d794

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7933/head:pull/7933` \
`$ git checkout pull/7933`

Update a local copy of the PR: \
`$ git checkout pull/7933` \
`$ git pull https://git.openjdk.java.net/jdk pull/7933/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7933`

View PR using the GUI difftool: \
`$ git pr show -t 7933`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7933.diff">https://git.openjdk.java.net/jdk/pull/7933.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/stringopts.cpp [hunk 1]
```diff
@@ -413,12 +376,13 @@ Node_List PhaseStringOpts::collect_toString_calls() {
       worklist.push(n);
     }
   }
-
+  uint encount = 0;
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/stringopts.cpp:379]: `encount` -> `encountered`

File: src/hotspot/share/opto/stringopts.cpp [hunk 2]
```diff
@@ -218,10 +185,7 @@ class StringConcat : public ResourceObj {
   void maybe_log_transform() {
     CompileLog* log = _stringopts->C->log();
     if (log != NULL) {
-      log->head("replace_string_concat arguments='%d' string_alloc='%d' multiple='%d'",
-                num_arguments(),
-                _string_alloc != NULL,
-                _multiple);
+      log->head("replace_string_concat arguments='%d'", num_arguments());
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/stringopts.cpp:188]: Why did you remove the additional information from the logging message?

**Review by TobiHartmann — CHANGES_REQUESTED** [review]

File: src/hotspot/share/opto/stringopts.cpp (hunk 2 above)
Comment (navyxliu) [reply src/hotspot/share/opto/stringopts.cpp:188]: they are only informative. they won't change behavior. should I keep them?

Comment (navyxliu) [reply src/hotspot/share/opto/stringopts.cpp]: hi, @TobiHartmann , 
I bring back multiple. _string_alloc is always NULL. there's no way to set this member data.  so I delete it.

**Comment by navyxliu** [comment]
hi, @TobiHartmann , 
could you take a look at this?

File: src/hotspot/share/opto/stringopts.cpp [hunk 3]
```diff
@@ -431,10 +400,34 @@ Node_List PhaseStringOpts::collect_toString_calls() {
...
+// The criterion is that the length of argument is easy to work out in this phrase.
+// It will drop complex cases such as Object.
+//
+// Since it walks along the receivers of fluent-chain, it will give up if the codeshape is
+// not "fluent" enough. eg.
+//   StringBuilder sb = new StringBuilder();
+//   sb.append("foo");
+//   sb.toString();
+//
+// The receiver of toString method is the result of Allocation Node(CheckedCastPP).
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/stringopts.cpp:428]:
```suggestion
// The receiver of toString method is the result of Allocation Node(CheckCastPP).
```

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Review by vnkozlov — APPROVED** [review]
Changes seem fine.

This code become almost obsolete after [JEP 280](https://openjdk.java.net/jeps/280)

You may need to test with `-XDstringConcat=inline` to exercise this C2's code more.

**Comment by navyxliu** [comment]
hi, @vnkozlov, @TobiHartmann 
Thank you for taking a look at this.
 
> This code become almost obsolete after [JEP 280](https://openjdk.java.net/jeps/280) 

I am actually quite confused about `StringOpts`.  After stepping into code detail, I realize the phase is designed for "String concatenation" produced by javac.  As you said, javac has changed its default concatenation policy in JEP-280. **Will we delete this phase in near future?**

IIUC, now PhaseStringOpts can improve code in 2 cases. The first case is the legacy bytecodes, or bytecodes generated by "-XDstringConcat=inline", like you hint. The second one is hand-writing StringBuilder.  For the second case, current implementation is incomplete.  I guess many Java developers (including me) don't realize that "fluent-chain" is mandatory here. if you accidentally break the fluent chain,  PhaseStringOpts will fail to recognize the pattern.   here I modify the microbenchmark from [Jose's.](https://github.com/JosePaumard/jep-cafe-07-string-concatenation/blob/master/src/main/java/org/paumard/jepcafe/StringConcat.java#L69)  and show the difference. 

```
Benchmark                            (converted)  (size)  Mode  Cnt    Score    Error  Units
StringConcat.stringBuilder4_2               1000      10  avgt   30  657.011 ± 18.321  ns/op
StringConcat.stringBuilder4_fluent2         1000      10  avgt   30   74.649 ±  1.085  ns/op

    @Benchmark
    public String stringBuilder4_2() {
        StringBuilder sb = new StringBuilder(1024);
        sb.append(s0).append(s1).append(s2).append(s3);
        return sb.toString();
    }

    @Benchmark
    public String stringBuilder4_fluent2() {
        StringBuilder sb = new StringBuilder(1024);
        return sb.append(s0).append(s1).append(s2).append(s3).toString();
    }
```

I have a patch to recognize the loose form, but I am not sure we should pursuit it or not. In real world,  it's very unlikely to use StringBuilder to concatenate some strings or integers, right?  "+" is the right way to go, right?

**Comment by vnkozlov** [comment]
First, lets push this changeset - it is reviewed and tested.
Second, `-XDstringConcat=inline` is used in JDK build system (`grep` it) so this code is still used.

If the patch for recognizing the loose form is not big we can go with it. As separate changes.

**Comment by navyxliu** [comment]
current testing macro does not support "-javacoptions" of jtreg.  it's difficult to cover all tests with -XDstringConcat=inline. however, tests in "test/jdk/java/lang/String/concat/" do use -XDstringConcat=inline to cover this it.  Also this change doesn't change any logic of transformation. I think it's safe.

**Comment by navyxliu** [comment]
I grep "-DstringConcat=inline" in make directory.  "stringConcate=inline" is indeed used to build javac! That explains why example above reports 79 encounters in javac.  thanks for the head-up
